core/src/test/scala/org/http4s/UriSpec.scala

```diff
@@ -578,4 +574,15 @@ class UriSpec extends Http4sSpec with MustThrownMatchers {
       u =? ps must be_==(u =? ps)
     }
   }
+
+  "Uri.renderString" should {
+    "Encode special chars in the query" in {
+      val u = Uri().withQueryParam("foo", " !$&'()*+,;=:/?@~")
+      u.renderString must_== "/?foo=%20%21%24%26%27%28%29%2A%2B%2C%3B%3D%3A%2F%3F%40~"
+    }
+    "Encode special chars in the fragment" in {
+      val u = Uri(fragment = Some(" !$&'()*+,;=:/?@~"))
+      u.renderString must_== "/#%20!$&'()*+,;=:/?@~"
+    }
```

Comment:
maybe we should add these tests in `Query` since `Uri`? 
If we have a method to extract the value from a `Query` it would be nice to check the round trip (e.g. Query.withQueryParam("foo", Seq(...)).get("foo"))